File: src/taeeun/week6/수식최대화.js
function solution(expression) {
    let answer = [];
    let op=/\-|\*|\+/g;
    //1. 숫자와 연산자를 각각 배열로 분리
    //숫자만 저장된 배열
    let numArr=expression.split(op).map(i=>parseInt(i));
    //연산자만 저장된 배열
    let opArr=[];
    for (let i=0; i<expression.length;i++){
        if (expression[i]==='-'||expression[i]==='+'||expression[i]==='*') opArr.push(expression[i]);
    }
    //2. 연산자 우선순위 정하기
    const priorities=[
        '*-+',
        '*+-',
        '+-*',
        '+*-',
        '-*+',
        '-+*',
    ];
    //3. 완전 탐색
    for (let i=0; i<priorities.length;i++){
        //새로운 배열에 복사
        let newNumArr=numArr.slice();
        let newOpArr=opArr.slice();
        for (let j=0; j<priorities[i].length; j++){
            let k=0;
            while (k!=newOpArr.length){
                if (priorities[i][j]===newOpArr[k]) {
                    if (newOpArr[k]==='*'){
                        newNumArr[k]*=newNumArr[k+1];
                    }
                    else if (newOpArr[k]==='-'){
                        newNumArr[k]-=newNumArr[k+1];
                    }
                    else if (newOpArr[k]==='+'){
                        newNumArr[k]+=newNumArr[k+1];
                    }
                    newNumArr.splice(k+1,1);
                    newOpArr.splice(k,1);
                }
                else k++;
            }
        }
        answer.push(newNumArr[0]);
    }
    return Math.max(...answer.map(i=>Math.abs(i)));
}

File: src/jimin/week6/수식 최대화.py
#수식 최대화
from itertools import permutations
from collections import deque
import re
from copy import deepcopy

def calculator(operator, num1, num2):
    if operator == '*':
        return num1 * num2
    elif operator == '-':
        return num1 - num2
    elif operator == '+':
        return num1 + num2

def solution(expression):
    answer = 0
    operators = ['*', '-', '+']

    # num = ""
    # op_list = deque()
    # for ex in expression:
    #     if ex.isdigit():
    #         num += ex
    #     else:
    #         op_list.append(int(num))
    #         op_list.append(ex)
    #         num = ""
    # op_list.append(int(num))

    op_list = re.split(r'(\D)', expression)
    op_list = deque([int(o) if o.isdigit() else o for o in op_list])

    for operator in list(permutations(operators, 3)):
        op_list_tmp = deepcopy(op_list)
        for o in operator: #* - + 연산자
            temp = deque()
            while len(op_list_tmp) != 0:
                curr = op_list_tmp.popleft()
                if curr == o:
                    temp.append(calculator(curr, temp.pop(), op_list_tmp.popleft()))
                else:
                    temp.append(curr)
            op_list_tmp = temp
        answer = max(answer, abs(op_list_tmp[-1]))

    return answer

expression = "100-200*300-500+20"
print(solution(expression))